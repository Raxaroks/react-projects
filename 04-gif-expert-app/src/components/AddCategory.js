import React, { useState } from 'react';
import PropTypes from 'prop-types';

const AddCategory = ( { setCategories } ) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (evt) => {
        setInputValue(evt.target.value);
        // console.log('handleInputChange invoked!!!');
    }

    const handleSubmit = (evt) => {
        evt.preventDefault();

        if ( inputValue.trim().length > 2 ) {
            setCategories( cats => [inputValue, ...cats] );
            setInputValue('');
        }
        // console.log('handleSubmit invoked!!!');
    }

    return (
        <form onSubmit={ handleSubmit }>
            <p>{inputValue}</p>
            <input
                type='text'
                placeholder='Escribe...' 
                value={ inputValue }
                onChange={ handleInputChange }
            />
        </form>
    );
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}

export default AddCategory;
