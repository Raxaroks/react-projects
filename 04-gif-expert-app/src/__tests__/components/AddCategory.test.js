import '@testing-library/jest-dom';
import React from 'react';
import { shallow } from 'enzyme';
import AddCategory from '../../components/AddCategory';

describe('Pruebas en el componente: <AddCategory />', () => {

    const setCategories = jest.fn(); // function mock
    let componentWrapper = shallow(<AddCategory setCategories={ setCategories }/>);

    beforeEach( () => {
        componentWrapper = shallow(<AddCategory setCategories={ setCategories }/>);
        jest.clearAllMocks();
    } )

    test('Debe de renderizarse y mostrarse correctamente.', () => {
        expect(componentWrapper).toMatchSnapshot();
    })

    test('Debe de cambiar el contenido del input', () => {        
        const input = componentWrapper.find('input');
        const value = 'Hola, Mundo!';

        input.simulate( 'change', {target: {value}} );
        expect( componentWrapper.find('p').text().trim() ).toBe(value);
    })

    test('No debe de postear la informaciÃ³n del formulario', () => {
        componentWrapper.find('form').simulate('submit', {preventDefault(){}});
        expect(setCategories).not.toHaveBeenCalled();
    })
    
    test('Debe de llamar a setCategories y limpiar el contenido del input', () => {
        const value = 'Un valor random juasjuas';
        componentWrapper.find('input').simulate('change', {target:{value}});
        componentWrapper.find('form').simulate('submit', {preventDefault(){}});;

        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));
        expect( componentWrapper.find('input').prop('value') ).toBe('');
    })
    
})
