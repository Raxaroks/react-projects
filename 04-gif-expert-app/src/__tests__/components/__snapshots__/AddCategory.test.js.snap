// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente: <AddCategory /> Debe de renderizarse y mostrarse correctamente. 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    placeholder="Escribe..."
    type="text"
    value=""
  />
</form>
`;
